feat(sync-service): Improve performance of LRU shape expiry

robacourt opened the pull request:
Turbo.ai have been seeing 2-8s latency on requests when the number of shapes exceeds the maximum number of shapes and LRU expiry kicks in. This is due to the shape cache process being blocked while the expiry happens, which can take a few seconds.

This PR moves expiry out of the ShapeCache process and into a new ExpiryManager process. When a new shape is added, the ShapeCache calls `ExpiryManager.notify_new_shape_added/1` which is non-blocking and debounced, and if max_shapes has been set, triggers finding the LRU shapes and expiring them.

balegas commented:
This sounds good. Deleting the files will be handled at `ExpiryShapeManager` speed and off-the-critical path.

alco commented on packages/sync-service/lib/electric/shape_cache/expiry_manager.ex:
```diff
@@ -0,0 +1,142 @@
+          number_to_expire: number_to_expire
+        ],
+        fn ->
+          shapes_to_expire
+          |> Enum.each(fn shape -> expire_shape(shape, state) end)
```
Is it safe to replace this with `Task.async_stream()`? It is worth expiring shapes in parallel with max concurrency set to e.g. 10-20.

robacourt replied:
will look at in another PR

alco reviewed (approved):
Lovely !

balegas commented:
I thought about that too, but I was not sure if we deleted shapes in any other way like 409s etc, that would not go through this code path

robacourt commented:
> * we want to add a bit of headroom whenever we clear shapes. So instead of bringing back the number of shapes to the max limit. We'll add a second variable which is the target number of shapes after a cleanup

I've added `ELECTRIC_EXPERIMENTAL_EXPIRY_RATIO` with a default of 0.2

> * In the telemetry we want to know the number of shapes we've cleaned up and the duration of a cleaning round. This will allow us to analyze if things are slow because they are slow, or just because we're deleting lots of shapes

yup, that's in there

> We had a discussion about we're sending a message to the ExpiryManager on every new shape to check if we need to cleanup vs. making the cleanup run periodically. There are trade-offs, but we expect that the number of new shapes not to be that large that would become an issue. It does seem that we should have a reasonable debouncing window for checking number of shapes if the counting of shapes is a bit compute intensive.

I've added a `recheck_delay_ms` of 1 second so the count will not be checked more than once a second. This is useful on top of debouncing when under the number of shapes is less than the max_shapes.
